department-of-veterans-affairs/developer-portal
# Validate Apply fields on blur, not on change

completes [API-154](https://vajira.max.gov/browse/API-154)

this PR restricts validation for the email and redirect URI fields to the blur event so that errors do not fire when the user is still typing. it also contains some refactoring of the post-submit state management, which i initially started doing because the form didn't clear after a successful submit.

couple of notes on `formation-react` things:

* the concept of "dirty" in text inputs is pretty much backwards. on the input's change event, it calls `onValueChange` with dirty set to the value of `props.field.dirty`, but the change event implies it _is_ dirty. on the input's blur event, it calls `onValueChange` with dirty set to true, but you can focus on the input and then move focus elsewhere without changing the value. so there's some lengthy comments to help explain that.
* we need to manually restore pre-existing validation on an `IErrorableInput` for a validated field, because the `formation-react` inputs don't handle that.

testing instructions:

* Apply form validation (do the following for both email and redirect URI)
  * start typing in a value that doesn't match the required criteria. ensure that you don't see an error message until after you leave the input.
  * return to the field and correct the error. ensure that the error message doesn't disappear until after you leave the input.
  * repeat the above steps with a screen reader. (tricky to do if you're not used to using the screen reader, which i assume y'all aren't. but it's good practice, and you can get a sense of what screen reader users are experience.)
* for the post-submit refactoring, try signing up with OAuth APIs only, standard APIs only, and a combination of both and ensure the sign-up works as expected.
  * navigate back to `/apply` and check that the form is blank.
